protocol-units/bridge/contracts/src/AtomicBridgeInitator.sol

```diff
@@ -0,0 +1,96 @@
+        bytes32 recipient;
+        bytes32 hashLock;
+        uint256 timeLock;
+        bool completed;
+    }
+
+    mapping(bytes32 => BridgeTransfer) public bridgeTransfers;
+    bytes32[] public bridgeTransferIds;
+    IWETH9 public weth;
+    uint256 private nonce;
+
+    function initialize(address _weth) public initializer {
+        if (_weth == address(0)) {
+            revert ZeroAddress();
+        }
+        weth = IWETH9(_weth);
+    }
+
+    function initiateBridgeTransfer(uint256 wethAmount, bytes32 recipient, bytes32 hashLock, uint256 timeLock)
+        external
+        payable
+        returns (bytes32 bridgeTransferId)
+    {
+        address originator = msg.sender;
+        uint256 ethAmount = msg.value;
+        uint256 totalAmount = wethAmount + ethAmount;
+        // Ensure there is a valid total amount
+        if (totalAmount == 0) {
+            revert ZeroAmount();
+        }
+        // If msg.value is greater than 0, convert ETH to WETH
+        if (ethAmount > 0) weth.deposit{value: ethAmount}();
+        //Transfer WETH to this contract, revert if transfer fails
+        if (wethAmount > 0) {
+            if (!weth.transferFrom(originator, address(this), wethAmount)) revert WETHTransferFailed();
+        }
+
+        nonce++; //increment the nonce
+        bridgeTransferId =
+            keccak256(abi.encodePacked(originator, recipient, hashLock, timeLock, block.timestamp, nonce));
+
+        // Check if the bridge transfer already exists
+        if (bridgeTransfers[bridgeTransferId].amount != 0) revert BridgeTransferInvalid();
+
+        bridgeTransfers[bridgeTransferId] = BridgeTransfer({
+            amount: totalAmount,
+            originator: originator,
+            recipient: recipient,
+            hashLock: hashLock,
+            timeLock: block.timestamp + timeLock,
+            completed: false
+        });
+
+        bridgeTransferIds.push(bridgeTransferId);
+
+        emit BridgeTransferInitiated(bridgeTransferId, originator, recipient, totalAmount, hashLock, timeLock);
+        return bridgeTransferId;
+    }
+
+    function completeBridgeTransfer(bytes32 bridgeTransferId, bytes32 preImage) external {
+        // Retrieve the bridge transfer
+        BridgeTransfer storage bridgeTransfer = bridgeTransfers[bridgeTransferId];
+        if (bridgeTransfer.completed) revert BridgeTransferHasBeenCompleted();
+        if (keccak256(abi.encodePacked(preImage)) != bridgeTransfer.hashLock) revert InvalidSecret();
+        bridgeTransfer.completed = true;
+        emit BridgeTransferCompleted(bridgeTransferId, preImage);
+    }
+
+    function refundBridgeTransfer(bytes32 bridgeTransferId) external {
+        BridgeTransfer storage bridgeTransfer = bridgeTransfers[bridgeTransferId];
+        uint256 amount = bridgeTransfer.amount;
+        if (bridgeTransfer.completed) revert BridgeTransferHasBeenCompleted();
+        if (block.timestamp < bridgeTransfer.timeLock) revert TimeLockNotExpired();
```

Comment:
We discussed this yesterdar right? That there would be no way to sync block time between the chains. Maybe I misunderstood but I thought we settled on block.timestamp as sufficient, if comparing with the move timestamp now() method.

Reply:
That is interesting. In my memory, we should look at actually, let's say, how time is actually secured on Ethereum. If that would be done by the validators, then we might do time, but also we get all this kind of time zone stuff going on. And thus we came to the conclusion that block height would be more absolute point in time because we know the moment the swap gets initiated what block height, what the height is of the block and we have an absolute number of blocks we will allow the bridge transfer to complete.

The other option would be kind of, let's say, to use an oracle, but then you're actually delegating the whole thing (responsibility) to oracles, right, which is again kind of an indirection. As such, we concluded, in my recollection, block height is probably the best way to go.

Reply:
Let's  just use block height on both sides because it's the most robust thing. We can script the logic inside the bridging service to take into account the average block time plus minus something and we take the range big enough. It is the most objective most consistent way after thinking about it. It will prevent all kinds of issues down the line.

Reply:
Ok that's done